File: src/FileUtilities.cpp
#include "FileUtilities.h"

#include <algorithm>
#include <fstream>
#include <stdexcept>

namespace fs = std::filesystem;

// ── Path helpers ──────────────────────────────────────────────────────────────

fs::path FileUtilities::join(const fs::path& base, const fs::path& relative) {
    return base / relative;
}

fs::path FileUtilities::normalize(const fs::path& path) {
    return fs::weakly_canonical(path);
}

bool FileUtilities::exists(const fs::path& path) {
    return fs::exists(path);
}

void FileUtilities::ensureDirectory(const fs::path& path) {
    std::error_code ec;
    fs::create_directories(path, ec);
    if (ec) {
        throw std::runtime_error("Failed to create directory '" + path.string() +
                                 "': " + ec.message());
    }
}

// ── Read / write ──────────────────────────────────────────────────────────────

std::string FileUtilities::readText(const fs::path& path) {
    std::ifstream file(path);
    if (!file) {
        throw std::runtime_error("Cannot open file for reading: " + path.string());
    }
    return {std::istreambuf_iterator<char>(file), std::istreambuf_iterator<char>()};
}

void FileUtilities::writeText(const fs::path& path,
                               const std::string& content,
                               bool overwrite) {
    if (!overwrite && fs::exists(path)) {
        throw std::runtime_error("File already exists (overwrite=false): " +
                                 path.string());
    }
    std::ofstream file(path);
    if (!file) {
        throw std::runtime_error("Cannot open file for writing: " + path.string());
    }
    file << content;
    if (!file) {
        throw std::runtime_error("Failed to write to file: " + path.string());
    }
}

std::vector<std::uint8_t> FileUtilities::readBytes(const fs::path& path) {
    std::ifstream file(path, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Cannot open file for reading: " + path.string());
    }
    return {std::istreambuf_iterator<char>(file), std::istreambuf_iterator<char>()};
}

void FileUtilities::writeBytes(const fs::path& path,
                                const std::vector<std::uint8_t>& data,
                                bool overwrite) {
    if (!overwrite && fs::exists(path)) {
        throw std::runtime_error("File already exists (overwrite=false): " +
                                 path.string());
    }
    std::ofstream file(path, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Cannot open file for writing: " + path.string());
    }
    file.write(reinterpret_cast<const char*>(data.data()),
               static_cast<std::streamsize>(data.size()));
    if (!file) {
        throw std::runtime_error("Failed to write to file: " + path.string());
    }
}

nlohmann::json FileUtilities::readJson(const fs::path& path) {
    const std::string text = readText(path);
    try {
        return nlohmann::json::parse(text);
    } catch (const nlohmann::json::parse_error& e) {
        throw std::runtime_error("JSON parse error in '" + path.string() +
                                 "': " + e.what());
    }
}

void FileUtilities::writeJson(const fs::path& path,
                               const nlohmann::json& json,
                               bool overwrite,
                               int indent) {
    writeText(path, json.dump(indent) + "\n", overwrite);
}

// ── Copy / move / delete ──────────────────────────────────────────────────────

void FileUtilities::copy(const fs::path& source,
                          const fs::path& dest,
                          bool overwrite) {
    const auto opts = overwrite ? fs::copy_options::overwrite_existing
                                : fs::copy_options::none;
    std::error_code ec;
    fs::copy(source, dest, opts, ec);
    if (ec) {
        throw std::runtime_error("Copy failed from '" + source.string() + "' to '" +
                                 dest.string() + "': " + ec.message());
    }
}

void FileUtilities::move(const fs::path& source,
                          const fs::path& dest,
                          bool overwrite) {
    if (!overwrite && fs::exists(dest)) {
        throw std::runtime_error("Destination already exists (overwrite=false): " +
                                 dest.string());
    }
    std::error_code ec;
    fs::rename(source, dest, ec);
    if (ec) {
        // fs::rename fails across devices; fall back to copy-then-delete
        fs::copy(source, dest, fs::copy_options::overwrite_existing, ec);
        if (ec) {
            throw std::runtime_error("Move (copy stage) failed from '" +
                                     source.string() + "': " + ec.message());
        }
        fs::remove(source, ec);
        if (ec) {
            throw std::runtime_error("Move (remove stage) failed for '" +
                                     source.string() + "': " + ec.message());
        }
    }
}

void FileUtilities::remove(const fs::path& path, bool missingOk) {
    if (!fs::exists(path)) {
        if (missingOk) {
            return;
        }
        throw std::runtime_error("File does not exist: " + path.string());
    }
    std::error_code ec;
    fs::remove_all(path, ec);
    if (ec) {
        throw std::runtime_error("Failed to remove '" + path.string() +
                                 "': " + ec.message());
    }
}

// ── Directory enumeration ─────────────────────────────────────────────────────

std::vector<fs::path> FileUtilities::listDirectory(const fs::path& dir) {
    if (!fs::is_directory(dir)) {
        throw std::runtime_error("Not a directory: " + dir.string());
    }
    std::vector<fs::path> entries;
    for (const auto& entry : fs::directory_iterator(dir)) {
        entries.push_back(entry.path());
    }
    std::sort(entries.begin(), entries.end());
    return entries;
}

namespace {

// Wildcard match supporting * (any sequence) and ? (any single character)
bool wildcardMatch(const std::string& pattern, const std::string& text) {
    const char* p = pattern.c_str();
    const char* t = text.c_str();
    const char* star = nullptr;
    const char* afterStar = nullptr;

    while (*t != '\0') {
        if (*p == '?' || *p == *t) {
            ++p;
            ++t;
        } else if (*p == '*') {
            star = p++;
            afterStar = t;
        } else if (star != nullptr) {
            p = star + 1;
            t = ++afterStar;
        } else {
            return false;
        }
    }
    while (*p == '*') {
        ++p;
    }
    return *p == '\0';
}

} // namespace

std::vector<fs::path> FileUtilities::glob(const fs::path& dir,
                                           const std::string& pattern) {
    if (!fs::is_directory(dir)) {
        throw std::runtime_error("Not a directory: " + dir.string());
    }
    std::vector<fs::path> results;
    for (const auto& entry : fs::directory_iterator(dir)) {
        if (wildcardMatch(pattern, entry.path().filename().string())) {
            results.push_back(entry.path());
        }
    }
    std::sort(results.begin(), results.end());
    return results;
}

std::vector<fs::path> FileUtilities::rglob(const fs::path& dir,
                                             const std::string& pattern) {
    if (!fs::is_directory(dir)) {
        throw std::runtime_error("Not a directory: " + dir.string());
    }
    std::vector<fs::path> results;
    for (const auto& entry : fs::recursive_directory_iterator(dir)) {
        if (wildcardMatch(pattern, entry.path().filename().string())) {
            results.push_back(entry.path());
        }
    }
    std::sort(results.begin(), results.end());
    return results;
}

std::vector<fs::path> FileUtilities::filterByExtension(
    const std::vector<fs::path>& files, const std::string& extension) {
    const std::string ext =
        (!extension.empty() && extension[0] != '.') ? ("." + extension) : extension;
    std::vector<fs::path> result;
    for (const auto& f : files) {
        if (f.extension() == ext) {
            result.push_back(f);
        }
    }
    return result;
}

File: file_utilities.py
import json
import logging
import shutil
from pathlib import Path
from typing import List, Optional, Union

logger = logging.getLogger(__name__)

PathLike = Union[str, Path]


class FileUtilities:
    """Centralized file operations: paths, read/write, copy/move/delete, and directory enumeration."""

    # ── Path helpers ──────────────────────────────────────────────────────────

    @staticmethod
    def join(*parts: PathLike) -> Path:
        if not parts:
            raise ValueError("join requires at least one path part")
        return Path(parts[0]).joinpath(*parts[1:])

    @staticmethod
    def normalize(path: PathLike) -> Path:
        return Path(path).resolve()

    @staticmethod
    def exists(path: PathLike) -> bool:
        return Path(path).exists()

    @staticmethod
    def ensure_directory(path: PathLike) -> Path:
        p = Path(path)
        p.mkdir(parents=True, exist_ok=True)
        logger.debug("Ensured directory: %s", p)
        return p

    # ── Read / write helpers ──────────────────────────────────────────────────

    @staticmethod
    def read_text(path: PathLike, encoding: str = "utf-8") -> str:
        p = Path(path)
        try:
            return p.read_text(encoding=encoding)
        except OSError as exc:
            logger.error("Failed to read text file %s: %s", p, exc)
            raise

    @staticmethod
    def write_text(path: PathLike, content: str, encoding: str = "utf-8") -> None:
        p = Path(path)
        try:
            p.write_text(content, encoding=encoding)
            logger.debug("Wrote text file: %s", p)
        except OSError as exc:
            logger.error("Failed to write text file %s: %s", p, exc)
            raise

    @staticmethod
    def read_bytes(path: PathLike) -> bytes:
        p = Path(path)
        try:
            return p.read_bytes()
        except OSError as exc:
            logger.error("Failed to read binary file %s: %s", p, exc)
            raise

    @staticmethod
    def write_bytes(path: PathLike, content: bytes) -> None:
        p = Path(path)
        try:
            p.write_bytes(content)
            logger.debug("Wrote binary file: %s", p)
        except OSError as exc:
            logger.error("Failed to write binary file %s: %s", p, exc)
            raise

    @staticmethod
    def read_json(path: PathLike) -> object:
        p = Path(path)
        try:
            return json.loads(p.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError) as exc:
            logger.error("Failed to read JSON file %s: %s", p, exc)
            raise

    @staticmethod
    def write_json(path: PathLike, data: object, indent: int = 2) -> None:
        p = Path(path)
        try:
            p.write_text(json.dumps(data, indent=indent), encoding="utf-8")
            logger.debug("Wrote JSON file: %s", p)
        except (OSError, TypeError) as exc:
            logger.error("Failed to write JSON file %s: %s", p, exc)
            raise

    # ── Copy / move / delete ──────────────────────────────────────────────────

    @staticmethod
    def copy(src: PathLike, dst: PathLike, overwrite: bool = False) -> Path:
        src_path, dst_path = Path(src), Path(dst)
        if dst_path.exists() and not overwrite:
            raise FileExistsError(f"Destination already exists: {dst_path}")
        try:
            shutil.copy2(src_path, dst_path)
            logger.debug("Copied %s -> %s", src_path, dst_path)
            return dst_path
        except OSError as exc:
            logger.error("Failed to copy %s -> %s: %s", src_path, dst_path, exc)
            raise

    @staticmethod
    def move(src: PathLike, dst: PathLike, overwrite: bool = False) -> Path:
        src_path, dst_path = Path(src), Path(dst)
        if dst_path.exists() and not overwrite:
            raise FileExistsError(f"Destination already exists: {dst_path}")
        try:
            shutil.move(str(src_path), str(dst_path))
            logger.debug("Moved %s -> %s", src_path, dst_path)
            return Path(dst_path)
        except OSError as exc:
            logger.error("Failed to move %s -> %s: %s", src_path, dst_path, exc)
            raise

    @staticmethod
    def delete(path: PathLike, missing_ok: bool = False) -> None:
        p = Path(path)
        try:
            if p.is_dir():
                shutil.rmtree(p)
                logger.debug("Deleted directory: %s", p)
            else:
                p.unlink(missing_ok=missing_ok)
                logger.debug("Deleted file: %s", p)
        except OSError as exc:
            logger.error("Failed to delete %s: %s", p, exc)
            raise

    # ── Enumerate / glob / filter ─────────────────────────────────────────────

    @staticmethod
    def list_directory(path: PathLike) -> List[Path]:
        p = Path(path)
        try:
            return sorted(p.iterdir())
        except OSError as exc:
            logger.error("Failed to list directory %s: %s", p, exc)
            raise

    @staticmethod
    def glob(path: PathLike, pattern: str) -> List[Path]:
        p = Path(path)
        try:
            return sorted(p.glob(pattern))
        except OSError as exc:
            logger.error("Failed to glob %s with pattern '%s': %s", p, pattern, exc)
            raise

    @staticmethod
    def rglob(path: PathLike, pattern: str) -> List[Path]:
        p = Path(path)
        try:
            return sorted(p.rglob(pattern))
        except OSError as exc:
            logger.error("Failed to rglob %s with pattern '%s': %s", p, pattern, exc)
            raise

    @staticmethod
    def filter_files(paths: List[Path], extension: Optional[str] = None) -> List[Path]:
        result = [p for p in paths if p.is_file()]
        if extension is not None:
            ext = extension if extension.startswith(".") else f".{extension}"
            result = [p for p in result if p.suffix == ext]
        return result
